### PR TITLE
CRM-20427 Fix issue with the regex in I18n query rewriting and add tests

### PR DESCRIPTION
**Pull request by seamuslee001**
* [CRM-20427: Many pages break on multilingual on 4.7.18](https://issues.civicrm.org/jira/browse/CRM-20427)

### CODE REVIEW

**Comment by seamuslee001** [comment]
@PalanteJon Jon can you test this out and see if it works for you?

**Comment by mlutfy** [comment]
Looks good (especially with a test!)

ping @samuelsov ?

File: tests/phpunit/CRM/Core/I18n/SchemaTest.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    $dbLocale = '_en_US';
+    $query = "Select * FROM {$table}";
+    $new_query = CRM_Core_I18n_Schema::rewriteQuery($query);
+    $this->assertEquals("Select * FROM {$expectedRewrite}", $new_query);
+    $query2 = "Select * FROM {$table} LIMIT 1";
+    $new_query2 = CRM_Core_I18n_Schema::rewriteQuery($query2);
+    $this->assertEquals("Select * FROM {$expectedRewrite} LIMIT 1", $new_query2);
+    $query3 = "SELECT * FROM {$table} JOIN civicrm_contact LIMIT 1";
+    $new_query3 = CRM_Core_I18n_Schema::rewriteQuery($query3);
+    $this->assertEquals("SELECT * FROM {$expectedRewrite} JOIN civicrm_contact LIMIT 1", $new_query3);
```
Comment (totten) [inline tests/phpunit/CRM/Core/I18n/SchemaTest.php]: @seamuslee001 Hooray for tests! :+1: 

I'd suggest adding some more edge-cases:

```sql
-- The `$table` appears in a string -- it should not be rewritten.
SELECT id FROM civicrm_activity WHERE subject = "civicrm_option_value"
SELECT "Fixed the the civicrm_option_value ticket"
SELECT 'Fixed the the civicrm_option_value ticket'

-- The table name is escaped. This is different from string-escaping, and it should be rewritten.
SELECT name FROM `civicrm_option_value` WHERE id = 123

-- The table name is part of a subquery
SELECT * FROM civicrm_foo WHERE foo_id = (SELECT value FROM civicrm_option_value)

-- Alternative verbs
DELETE FROM civicrm_option_value
INSERT INTO civicrm_option_value (foo, bar) VALUES (123, "Just a civicrm_option_value string")
```

**Comment by eileenmcnaughton** [comment]
Damn @seamuslee001 can we clone you!

**Comment by seamuslee001** [comment]
I have now added some more edge cases as identified by Tim

However it seems to be failing on the SELECT "This is a quoted string with <tablename> Tickets" Where it is wanting to localise the table name. 

@mlutfy @PalanteJon @samuelsov @totten should that be happening or not? and Also i'm not great with regex and this is what i came up with after some quick googling

**Comment by totten** [comment]
I think the test is correctly reporting an error in the scenario where a string includes the table name. Ex: Create an `Event` with `title`=`Celebrate 10th anniversary of civicrm_option_value table` and search for it using the API:

```
## This works as expected
cv ev 'return civicrm_api3("event","get",["title"=>["LIKE"=>"%Celebrate%"]]);'

## This doesn't work as expected... due to the query-rewriting
cv ev 'return civicrm_api3("event","get",["title"=>["LIKE"=>"%civicrm_option_value%"]]);'
```

I don't know how to fix that with a regex tweak -- some kind of SQL tokenizer or parser seems more appropriate. However, it's fair to say that this bug is pre-existing and uncommon and shouldn't be a blocker on fixing CRM-20427.

**Comment by seamuslee001** [comment]
I think this is s a pretty tiny edge case, I think if Jon confirms that this fixes the problem he was facing, Then i would say we comment out the tests that are failing at the moment indicating they should be fixed and work on a separate ticket pr to address that issue

**Comment by eileenmcnaughton** [comment]
sounds fair - unless @totten thinks he has the magic regex

**Comment by samuelsov** [comment]
@seamuslee001 thanks for the fast resolution and unit testing. It works in every case i have tested. The edge cases you and totten mention seems to be really minor problems imo and could be hard to resolve (or leading to potential regression) so i'd prefer ignore them.

**Comment by eileenmcnaughton** [comment]
OK - seems OK to merge to me

**Comment by seamuslee001** [comment]
On the back of @samuelsov 's comment i have altered tests to skip the ones that are failing as they are really edge cases.

**Comment by PalanteJon** [comment]
OK, I finally had a chance to test this and can confirm it fixes the bug I described in CRM-20427.

**Pull request merged**